sway/config/bar.c

```diff
@@ -188,70 +188,14 @@ static void handle_swaybar_client_destroy(struct wl_listener *listener,
 }
 
 static void invoke_swaybar(struct bar_config *bar) {
-	int sockets[2];
-	if (socketpair(AF_UNIX, SOCK_STREAM, 0, sockets) != 0) {
-		sway_log_errno(SWAY_ERROR, "socketpair failed");
-		return;
-	}
-	if (!set_cloexec(sockets[0], true) || !set_cloexec(sockets[1], true)) {
-		return;
-	}
-
-	bar->client = wl_client_create(server.wl_display, sockets[0]);
-	if (bar->client == NULL) {
-		sway_log_errno(SWAY_ERROR, "wl_client_create failed");
-		return;
-	}
-
+	char *const cmd[] = {
+			bar->swaybar_command ? bar->swaybar_command : "swaybar",
+			"-b", bar->id, NULL};
 	bar->client_destroy.notify = handle_swaybar_client_destroy;
-	wl_client_add_destroy_listener(bar->client, &bar->client_destroy);
-
-	pid_t pid = fork();
-	if (pid < 0) {
-		sway_log(SWAY_ERROR, "Failed to create fork for swaybar");
-		return;
-	} else if (pid == 0) {
-		// Remove the SIGUSR1 handler that wlroots adds for xwayland
-		sigset_t set;
-		sigemptyset(&set);
-		sigprocmask(SIG_SETMASK, &set, NULL);
-
-		pid = fork();
-		if (pid < 0) {
-			sway_log_errno(SWAY_ERROR, "fork failed");
-			_exit(EXIT_FAILURE);
-		} else if (pid == 0) {
-			if (!set_cloexec(sockets[1], false)) {
-				_exit(EXIT_FAILURE);
-			}
-
-			char wayland_socket_str[16];
-			snprintf(wayland_socket_str, sizeof(wayland_socket_str),
-					"%d", sockets[1]);
-			setenv("WAYLAND_SOCKET", wayland_socket_str, true);
-
-			// run custom swaybar
-			char *const cmd[] = {
-					bar->swaybar_command ? bar->swaybar_command : "swaybar",
-					"-b", bar->id, NULL};
-			execvp(cmd[0], cmd);
-			_exit(EXIT_FAILURE);
-		}
-		_exit(EXIT_SUCCESS);
-	}
-
-	if (close(sockets[1]) != 0) {
-		sway_log_errno(SWAY_ERROR, "close failed");
-		return;
+	bar->client = spawn_wl_client(cmd, server.wl_display, &bar->client_destroy);
```

Comment:
TBH I'd prefer setting up the listener manually in the caller.

Reply:
(This would also fix the listener being called in case of error)

Reply:
I don't know wayland internals but in most cases it's better to add handler as early as possible.
If it's not needed to be called in case of error it may be added before return from function.
Also it's possible to check this argument for NULL and add only if it's set

Reply:
I don't want our functions to look like this:

```c
struct thing *create_thing(struct wl_listener *event1_listener, struct wl_listener *event2_listener, …, struct wl_listener *eventN_listener);
```

Reply:
(And I don't want the listener to be added in case of error inside of `spawn_wl_client` -- right now it is)

Reply:
See fixup commit